extensions/amp-ad/0.1/amp-ad.js

```diff
@@ -0,0 +1,401 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {BaseElement} from '../../../src/base-element';
+import {IntersectionObserver} from '../../../src/intersection-observer';
+import {cidForOrNull} from '../../../src/cid';
+import {getIframe, prefetchBootstrap} from '../../../src/3p-frame';
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {listen, listenOnce, postMessage} from '../../../src/iframe-helper';
+import {loadPromise} from '../../../src/event-helper';
+import {parseUrl} from '../../../src/url';
+import {registerElement} from '../../../src/custom-element';
+import {adPrefetch, adPreconnect, clientIdScope} from '../../../ads/_config';
+import {timer} from '../../../src/timer';
+import {user} from '../../../src/log';
+import {userNotificationManagerFor} from '../../../src/user-notification';
+import {viewerFor} from '../../../src/viewer';
+import {installEmbed} from './amp-embed';
+
+/** @private @const These tags are allowed to have fixed positioning */
+const POSITION_FIXED_TAG_WHITELIST = {
+  'AMP-LIGHTBOX': true,
+};
+
+
+/**
+ * @type {boolean} Heuristic boolean as for whether another ad is currently
+ *     loading.
+ */
+let loadingAdsCount = 0;
+
+export class AmpAd extends BaseElement {
```

Comment:
For somewhat legacy reasons, extensions extend `AMP.BaseElement` instead. That avoids pulling the base class into the extension JS file.